Metronome accent patterns

pljones opened the pull request:
I'm raising this as a PR.

I still can't get the VST working.

When I load it into Reaper, it starts up okay and then shows this popup:
![jamtaba-launch-popup](https://user-images.githubusercontent.com/1549463/30607717-e652889a-9d6d-11e7-9463-4bf5b1575f8e.PNG)
After clicking OK, it just vanishes, still:
![jamtaba-launch-clicked-ok](https://user-images.githubusercontent.com/1549463/30607758-05132b2c-9d6e-11e7-85aa-e5858567a3fe.PNG)

Here's the debug output:
[JamTaba2.1.0-met.txt](https://github.com/elieserdejesus/JamTaba/files/1315384/JamTaba2.1.0-met.txt)

elieserdejesus commented:
@pljones , try run **qmake** in QtCreator (click with right button in project name) to force a recompilation.   
 I'm suspecting your Configurator.h (the VERSION) is not updated.

pljones commented:
Still getting the version error.  I have:
- deleted my local git checkout and cloned again
- deleted the output folder
- checked Configurator.h (says 2.1.0)
- run a Clean all
- run a qmake
- run a Rebuild all

jonjamcam commented:
@pljones did you uninstall other jamtaba vst from your system. It could be recognizing an old version.

pljones commented:
OK, uninstalling my day-to-day copy worked (I'd not upgraded for a while).

Previously this hasn't been an issue - I've been able to have an installed copy and my development copy has taken priority because it's later in the search path.

Still got the issue that I can't type space into the input field and I'm not getting change events from it, though.

elieserdejesus commented:
> Basically, I'm having no luck at all with the UI, any help there would be appreciated.

I will check soon. The spacebar issue is related with VST, I need check the details with more time.

elieserdejesus commented on src/Common/gui/NinjamPanel.cpp:
```diff
@@ -256,9 +258,34 @@ int NinjamPanel::getPanSliderMaximumValue() const
+void NinjamPanel::setAccentBeatsText(QString value)
+{
+    qCDebug(jtNinjamGUI) << "NinjamPanel::setCurrentAccentBeats " << value;
+    ui->lineEditAccentBeats->blockSignals(false);
```
@pljones , this is this problem. The block/unblock signals commands are inverted. Block first, and unblock in the end. Tested here and works fine after fix the block/unblock.

pljones replied:
You're right - but that didn't fix the space bar issue (still goes to the host).

elieserdejesus commented:
@pljones , lineEdits are *hacked* in VST plugin. The steps to get the accents line edit working are:

***
1 - Add a new parameter in NinjamPanel constructor:
![image](https://user-images.githubusercontent.com/1012741/30783189-0cf60698-a115-11e7-95fc-2910d1b8ef4d.png)

***

2 - Change the NinjamRoomWindow accordinly:

![image](https://user-images.githubusercontent.com/1012741/30783214-6f2fd870-a115-11e7-991b-6ee6c67ce686.png)

***

3 - Add a 3rd parameter in **NinjamPanel::initializeCombos** and modify (install the modifier) in your lineEdit.

![image](https://user-images.githubusercontent.com/1012741/30783230-ae205d8e-a115-11e7-9b27-723f69176d3f.png)

pljones commented:
Ah... I was wondering what the bpm/bpi code was doing.  I'll give that a go, thanks!

pljones commented:
Right, that's better.  However...
- the field is enabled even when "Custom..." isn't selected - that's not intended
- I clearly need to add a signal handler to catch the enter-completion

Still, it's much easier to test it now! 👍

elieserdejesus commented:
@pljones 
> the field is enabled even when "Custom..." isn't selected - that's not intended

Please replace the lineEdit **setEnabled** with **setVisible**, the lineEdit will appear only to the **custom** option. The UI will be more clean.

Another UI improvement is use the  QLineEdit **placeHolderText** to explain the accents pattern.

pljones commented:
OK, I'll to the visible as well - it will mean a gap in the UI when it's off, though.

pljones commented:
Hm, I've broken something.  I had it so enter was updating and now you need to switch off and on again.

elieserdejesus commented:
> The placeholder text doesn't appear to have any effect (nor does the tooltip - not that the tooltips work on other fields, either).

I suspect this is a VST only behavior, possible the textEdit hack is not considering the placehoder text and the tooltip.

pljones commented:
I'll see what I can do (once I get a build environment for the standalone...).

pljones commented:
Hi,

The latest on this is I'm stuck - as usual - with 64 bit portaudio not working with my ASIO drivers: it crashes on start up with no diagnostics.  (NOT a Jamtaba-specific issue as far as I'm aware.)

I didn't manage to get a build of portaudio I could use with Jamtaba, though.  If you could let me know how you build it, that could help.

Alternatively, just patch what's here now and I'll pull that back to my fork.

Thanks,

-- Peter

elieserdejesus commented:
> I didn't manage to get a build of portaudio I could use with Jamtaba, though. If you could let me know how you build it, that could help.

@pljones , I compiled portaudio long time ago. But I think I changed just 2 things in the Visual Studio project (it's inside **build** folder). 

1 - I setted the **PA_USE_ASIO** preprocessor macro in project properties.
2 - Changed the link options to compile as **MD** (if I remember correctly the original project is compiling as MT). All JamTAba libs are compiled with **MD** flag and I'm packaging the **vc redistributable**.

***

> Alternatively, just patch what's here now and I'll pull that back to my fork.

Ok, I will integrate your code, thanks for your contribution again.
